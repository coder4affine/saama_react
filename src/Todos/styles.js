const styles = {
  display: {
    display: 'flex',
  },
  flex: {
    flex: 1,
  },
  row: {
    display: 'flex',
    flexDirection: 'column',
  },
  vCenter: {
    justifyContent: 'center',
  },
  hCenter: {
    alignItems: 'center',
  },
  center: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  innerContainer: {
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between  ',
    flex: 1,
  },
  margin: {
    margin: 10,
  },
};

export default styles;
