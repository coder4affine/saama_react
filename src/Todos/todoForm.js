import React, { memo } from 'react';
import PropTypes from 'prop-types';

const todoForm = ({ submit, inputRef }) => (
  // console.log('render todoForm');
  // throw new Error('Something went wrong!');
  <form onSubmit={submit}>
    <input type="text" ref={inputRef} />
    <button type="submit">Add Todo</button>
  </form>
);
todoForm.propTypes = {
  submit: PropTypes.func.isRequired,
  inputRef: PropTypes.func.isRequired,
};

export default memo(todoForm);
