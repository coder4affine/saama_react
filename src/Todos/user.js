import React from 'react';
import PropTypes from 'prop-types';

const User = ({ user }) => <div>{user.name}</div>;

User.propTypes = {
  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }).isRequired,
};

export default User;
