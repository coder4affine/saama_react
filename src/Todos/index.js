import React, { useState, useRef, useEffect, memo } from 'react';
import TodoForm from './todoForm';
import TodoList from './todoList';
import TodoActions from './todoActions';
// import PropTypes from 'prop-types';
import styles from './styles';
import { statusText } from '../constants';
// import User from './user';

// class -> extend component/PureComponent(cant use shouldComponent Life cycle method)
// -> data manupulate / add logic
// -> state
// -> life cycle

// -> if youn want to use following life cycle method use class component
// -> getSnapshotBeforeUpdate
// -> componentDidCatch
// -> getDerivedStateFromError

// performace impacting lcm
// -> shouldComponentUpdate / PureComponent
// -> componentWillUnmount

// function -> with memo/ without memo
// -> Only For UI (props base)
// -> Always Use Memo

// mounting
// -> constructor
// -> getDerivedStateFromProps
// -> render
// -> componentDidMount

// update
// -> getDerivedStateFrmProps
// -> shouldComponentUpdate
// -> render
// -> getSnapshotBeforeUpdate
// -> componentDidUpdate

// unmount
// -> componentWillUnmount

// error
// -> componentDidCatch
// -> getDerivedStateFromError

// hello data
const Index = () => {
  const [todoList, setTodoList] = useState([]);
  const [status, setStatus] = useState(statusText[0]);
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(false);
  // const [temp, setTemp] = useState(`${data}, Yagnesh`);

  const inputText = useRef(null);

  const apiCall = (url, reqInit = {}) => {
    setLoading(true);
    return new Promise((resolve, reject) => {
      fetch(url, reqInit)
        .then((res) => res.json())
        .then((data) => {
          resolve(data);
        })
        .catch((err) => {
          setError(err);
          reject(err);
        })
        .finally(() => {
          setLoading(false);
        });
    });
  };

  const loadData = async () => {
    const data = await apiCall('http://localhost:3004/todos');
    setTodoList(data);
  };

  // componentDidMount
  useEffect(() => {
    loadData();
  }, []);

  // // componentDidUpdate
  // useEffect(() => {
  //   const timer = setTimeout(() => {
  //     console.log('timer');
  //   }, 1000);

  //   console.log('useEffect', loading);
  //   return () => {
  //     console.log('unmount');
  //     clearTimeout(timer);
  //   };
  // }, [loading, error]);

  // componentWillUnmount

  const addTodo = async (data) => {
    const newTodo = await apiCall('http://localhost:3004/todos', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        accept: 'application/json',
      },
      body: JSON.stringify(data),
    });
    console.warn(newTodo);
    setTodoList([newTodo, ...todoList]);
  };

  const updateTodo = async (data) => {
    const updatedTodo = await apiCall(`http://localhost:3004/todos/${data.id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        accept: 'application/json',
      },
      body: JSON.stringify(data),
    });
    const updatedTodoList = todoList.map((x) => {
      if (x.id === updatedTodo.id) {
        return updatedTodo;
      }
      return x;
    });
    setTodoList(updatedTodoList);
  };

  const deleteTodo = async (id) => {
    await apiCall(`http://localhost:3004/todos/${id}`, {
      method: 'DELETE',
    });
    setTodoList(todoList.filter((x) => x.id !== id));
  };

  const submit = async (e) => {
    e.preventDefault();
    const { value } = inputText.current;
    if (value) {
      await addTodo({ text: value, isDone: false });
      // this.todoText.value = '';
    }
  };

  const completeTodo = (todo) => {
    updateTodo({ ...todo, isDone: !todo.isDone });
    // [
    //     ...todoList.slice(0, i),
    //     { ...todoList[i], isDone: !todoList[i].isDone },
    //     ...todoList.slice(i + 1),
    //   ]
  };

  const changeStatus = (data) => {
    setStatus(data);
  };

  if (error) {
    return <h2 style={{ color: 'red' }}>{error.message}</h2>;
  }

  if (loading) {
    return <h3 style={{ color: 'blue' }}>Loading...</h3>;
  }

  return (
    <div
      style={{
        ...styles.row,
        ...styles.hCenter,
        ...styles.flex,
      }}
    >
      {/* <h1>{user.name}</h1> */}
      <h3 id="h3">To-Do App</h3>
      {/* <User user={user} />

      <button
        type="button"
        onClick={() => {
          this.setState({
            user: { ...user, name: 'rohit' },
          });
        }}
      >
        Change name
      </button> */}
      <TodoForm submit={submit} inputRef={inputText} />
      <div style={styles.innerContainer}>
        <TodoList
          todoList={todoList}
          status={status}
          completeTodo={completeTodo}
          deleteTodo={deleteTodo}
        />
        <TodoActions changeStatus={changeStatus} />
      </div>
    </div>
  );
};

export default memo(Index);

// class index extends Component {
//   constructor(props) {
//     super(props);
//     this.state = {
//       todoList: [],
//       status: statusText[0],
//       loading: false,
//       error: false,
//     };
//     // this.loadData();
//   }

//   componentDidMount() {
//     this.loadData();
//   }

//   componentDidCatch(error, info) {
//     console.warn(error);
//     console.warn(info);
//   }

//   static getDerivedStateFromError(error) {
//     return { error: error.message };
//   }

//   apiCall = (url, reqInit = {}) => {
//     this.setState({ loading: true });
//     return new Promise((resolve, reject) => {
//       fetch(url, reqInit)
//         .then((res) => res.json())
//         .then((data) => {
//           resolve(data);
//         })
//         .catch((error) => {
//           reject(error);
//         })
//         .finally(() => {
//           this.setState({ loading: false });
//         });
//     });
//   };

//   loadData = async () => {
//     const todoList = await this.apiCall('http://localhost:3004/todos');
//     this.setState({ todoList });
//   };

//   addTodo = async (data) => {
//     const newTodo = await this.apiCall('http://localhost:3004/todos', {
//       method: 'POST',
//       headers: {
//         'Content-Type': 'application/json',
//         accept: 'application/json',
//       },
//       body: JSON.stringify(data),
//     });
//     console.warn(newTodo);
//     const { todoList } = this.state;
//     this.setState({ todoList: [newTodo, ...todoList] }, () => {
//       // console.log(this.todoText);
//     });
//   };

//   updateTodo = async (data) => {
//     const updatedTodo = await this.apiCall(`http://localhost:3004/todos/${data.id}`, {
//       method: 'PUT',
//       headers: {
//         'Content-Type': 'application/json',
//         accept: 'application/json',
//       },
//       body: JSON.stringify(data),
//     });
//     const { todoList } = this.state;
//     const updatedTodoList = todoList.map((x) => {
//       if (x.id === updatedTodo.id) {
//         return updatedTodo;
//       }
//       return x;
//     });
//     this.setState({ todoList: updatedTodoList });
//   };

//   deleteTodo = async (id) => {
//     await this.apiCall(`http://localhost:3004/todos/${id}`, {
//       method: 'DELETE',
//     });
//     const { todoList } = this.state;
//     this.setState({ todoList: todoList.filter((x) => x.id !== id) });
//   };

//   submit = async (e) => {
//     e.preventDefault();
//     const { value } = this.todoText;
//     if (value) {
//       await this.addTodo({ text: value, isDone: false });
//       // this.todoText.value = '';
//     }
//   };

//   completeTodo = (todo) => {
//     this.updateTodo({ ...todo, isDone: !todo.isDone });
//     // [
//     //     ...todoList.slice(0, i),
//     //     { ...todoList[i], isDone: !todoList[i].isDone },
//     //     ...todoList.slice(i + 1),
//     //   ]
//   };

//   changeStatus = (status) => {
//     this.setState({ status });
//   };

//   render() {
//     const { todoList, status, error, loading } = this.state;
//     if (error) {
//       return <h2 style={{ color: 'red' }}>{error}</h2>;
//     }

//     if (loading) {
//       return <h3 style={{ color: 'blue' }}>Loading...</h3>;
//     }

//     return (
//       <div
//         style={{
//           ...styles.row,
//           ...styles.hCenter,
//           ...styles.flex,
//         }}
//       >
//         {/* <h1>{user.name}</h1> */}
//         <h3 id="h3">To-Do App</h3>
//         {/* <User user={user} />

//         <button
//           type="button"
//           onClick={() => {
//             this.setState({
//               user: { ...user, name: 'rohit' },
//             });
//           }}
//         >
//           Change name
//         </button> */}
//         <TodoForm
//           submit={this.submit}
//           inputRef={(ref) => {
//             this.todoText = ref;
//           }}
//         />
//         <div style={styles.innerContainer}>
//           <TodoList
//             todoList={todoList}
//             status={status}
//             completeTodo={this.completeTodo}
//             deleteTodo={this.deleteTodo}
//           />
//           <TodoActions changeStatus={this.changeStatus} />
//         </div>
//       </div>
//     );
//   }
// }

// index.propTypes = {};

// export default index;
