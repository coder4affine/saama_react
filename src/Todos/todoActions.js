import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { statusText } from '../constants';
import styles from './styles';

const todoActions = ({ changeStatus }) => {
  console.log('render todoActions');
  return (
    <div style={styles.display}>
      <button
        style={{ ...styles.flex, ...styles.center }}
        type="button"
        onClick={() => changeStatus(statusText[0])}
      >
        All
      </button>
      <button
        style={{ ...styles.flex, ...styles.center }}
        type="button"
        onClick={() => changeStatus(statusText[1])}
      >
        Pending
      </button>
      <button
        style={{ ...styles.flex, ...styles.center }}
        type="button"
        onClick={() => changeStatus(statusText[2])}
      >
        Completed
      </button>
    </div>
  );
};

todoActions.propTypes = {
  changeStatus: PropTypes.func.isRequired,
};

export default memo(todoActions);
