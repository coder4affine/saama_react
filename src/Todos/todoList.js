import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { statusText } from '../constants';
import styles from './styles';

const TodoList = ({ todoList, status, completeTodo, deleteTodo }) => {
  console.log('render TodoList');
  return (
    <div style={styles.flex}>
      {todoList
        .filter((x) => {
          if (status === statusText[1]) {
            return !x.isDone;
          }
          if (status === statusText[2]) {
            return x.isDone;
          }
          return true;
        })
        .map((todo) => (
          <div
            key={todo.id}
            style={{
              ...styles.display,
              ...styles.margin,
            }}
          >
            <input type="checkbox" checked={todo.isDone} onChange={() => completeTodo(todo)} />
            <span
              style={{
                ...styles.flex,
                textDecoration: todo.isDone ? 'line-through' : 'none',
              }}
            >
              {todo.text}
            </span>
            <button type="button" onClick={() => deleteTodo(todo.id)}>
              Delete
            </button>
          </div>
        ))}
    </div>
  );
};

TodoList.propTypes = {
  todoList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
      isDone: PropTypes.bool.isRequired,
    }),
  ).isRequired,
  status: PropTypes.string.isRequired,
  completeTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
};

export default memo(TodoList);
