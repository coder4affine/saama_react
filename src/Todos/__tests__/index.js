import 'isomorphic-fetch';
import React from 'react';
import { render, waitFor } from '@testing-library/react';
import Todos from '../index';

describe('Todos', () => {
  test('should render component', () => {
    const { container } = render(<Todos />);
    expect(container).toMatchSnapshot();
  });

  test('should render component after api call', async () => {
    const mockSuccessResponse = [
      {
        text: 'add new todo',
        isDone: true,
        id: 14,
      },
    ];

    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });

    const mockAPI = jest.spyOn(global, 'fetch').mockImplementationOnce(() => mockFetchPromise);

    const { container } = render(<Todos />);

    await waitFor(() => expect(mockAPI).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });
});
