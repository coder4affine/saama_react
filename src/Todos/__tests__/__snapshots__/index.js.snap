// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos should render component 1`] = `
<div>
  <h3
    style="color: blue;"
  >
    Loading...
  </h3>
</div>
`;

exports[`Todos should render component after api call 1`] = `
<div>
  <div
    style="display: flex; flex-direction: column; align-items: center; flex: 1;"
  >
    <h3
      id="h3"
    >
      To-Do App
    </h3>
    <form>
      <input
        type="text"
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
    <div
      style="width: 100%; display: flex; flex-direction: column; justify-content: space-between; flex: 1;"
    >
      <div
        style="flex: 1;"
      >
        <div
          style="display: flex; margin: 10px;"
        >
          <input
            checked=""
            type="checkbox"
          />
          <span
            style="flex: 1; text-decoration: line-through;"
          >
            add new todo
          </span>
          <button
            type="button"
          >
            Delete
          </button>
        </div>
      </div>
      <div
        style="display: flex;"
      >
        <button
          data-testid="btn-all"
          style="flex: 1; justify-content: center; align-items: center;"
          type="button"
        >
          All
        </button>
        <button
          data-testid="btn-pending"
          style="flex: 1; justify-content: center; align-items: center;"
          type="button"
        >
          Pending
        </button>
        <button
          data-testid="btn-completed"
          style="flex: 1; justify-content: center; align-items: center;"
          type="button"
        >
          Completed
        </button>
      </div>
    </div>
  </div>
</div>
`;
